src/patchkernel/patch_kernel_parallel.cpp

```diff
@@ -1750,15 +1657,231 @@ double PatchKernel::evalPartitioningUnbalance(const std::unordered_map<long, dou
 */
 std::vector<adaption::Info> PatchKernel::_partitioningPrepare(const std::unordered_map<long, double> &cellWeights, double defaultWeight, bool trackPartitioning)
 {
-	BITPIT_UNUSED(cellWeights);
-	BITPIT_UNUSED(defaultWeight);
-	BITPIT_UNUSED(trackPartitioning);
+	// Default partitioning can only be used when the patch is all on a single process
+	if (isDistributed()) {
+		throw std::runtime_error("Default partitioning can only be used when the patch is all on a single process");
+	}
+
+	// Early return if the mesh is empty
+	if (empty(true)) {
+		return std::vector<adaption::Info>();
+	}
+
+	// Early return if there is a single partition
+	if (getProcessorCount() == 1) {
+		return std::vector<adaption::Info>();
+	}
+
+	// Evaluate partitioning
+#if BITPIT_ENABLE_METIS==1
+	std::unordered_map<long, int> cellRanks;
+	if (getRank() == getOwner()) {
+		// Initialize map between patch vertices and METIS nodes
+		//
+		// METIS node numbering needs to be contiguous and needs to start from zero.
+		std::unordered_map<long, idx_t> vertexToNodeMap;
+
+		idx_t nodeId = 0;
+		for (const Vertex &vertex : m_vertices) {
+			vertexToNodeMap[vertex.getId()] = nodeId;
+			++nodeId;
+		}
+
+		// Create METIS mesh
+		idx_t nElements = getInternalCellCount();
+		idx_t nNodes = getInternalVertexCount();
+
+		idx_t connectSize = 0;
+		for (const Cell &cell : m_cells){
+			connectSize += static_cast<idx_t>(cell.getVertexCount());
+		}
+
+		std::vector<idx_t> connectStorage(connectSize);
+		std::vector<idx_t> connectRanges(nElements + 1);
+		std::vector<idx_t> elementWeights(nElements, static_cast<idx_t>(std::max(defaultWeight, 1.)));
+
+		idx_t i = 0;
+		idx_t j = 0;
+		connectRanges[0] = 0;
+		for (const Cell &cell : m_cells) {
+			// Element connectivity
+			for (long vertex : cell.getVertexIds()) {
+				connectStorage[j] = vertexToNodeMap[vertex];
+				++j;
+			}
+			connectRanges[i + 1] = j;
+
+			// Element weight
+			auto weightItr = cellWeights.find(cell.getId());
+			if (weightItr != cellWeights.end()) {
+				elementWeights[i] = static_cast<idx_t>(std::max(std::round(weightItr->second), 1.));
+			}
+
+			// Increase counters
+			i++;
+		}
+
+		// Get the number of common nodes two elements should share to be considered neighbours
+		idx_t nCommonNodes = getDimension();
+
+		// Get the number of partitions the mesh should be split into
+		idx_t nPartitions = getProcessorCount();
+
+		// Evaluate partitioning
+		idx_t objectiveValue;
+
+		std::vector<idx_t> elementRanks(nElements);
+		std::vector<idx_t> nodeRanks(nNodes);
+
+		int status = METIS_PartMeshDual(&nElements, &nNodes, connectRanges.data(), connectStorage.data(),
+										elementWeights.data(), nullptr, &nCommonNodes, &nPartitions, nullptr,
+										nullptr, &objectiveValue, elementRanks.data(), nodeRanks.data());
+
+		if (status != METIS_OK) {
+			throw std::runtime_error("Error during partitioning (METIS error " + std::to_string(status) + ")");
+		}
+
```

Comment:
Sorry for the dumb question,  but I don't know very well metis and just to foresee eventual bottlenecks in case I want to use directly this partitionPrepare Method with  cellweights. 
@edoardolombardi mimic used we used an external partitioner METIS_PartGraphKWay and we call a PatchKernel partition method employing a list of cellRanks to achieve distribution on ranks.  This bitpit version embeds a call to METIS_PartMeshDual instead, in the twin PatchKernel partition method using the cellWeights. There are substanstial differences between the 2? Advantages in using one instead of another?

Reply:
In mimic we use directly METIS_PartGraphKWay because we reconstruct the graph through the adjacencies structure of the patch. Inside METIS_PartMeshDual, I think it is done exactly the same (in a mesh-to-graph method, by supposing that two elements are neighbour when at least one edge is shared). 
In my opinion, the call to METIS_PartMeshDual can be useful if the adjacencies structure of the patch can be empty (@andrea-iob is it the case? What about the ghost layer/s creation during partition method?).
If here the adjacencies are always needed, the METIS_PartGraphKWay can be called directly and an internal double computation af tha adjacencies can be avoided.

Reply:
With METIS_PartMeshDual it's easier to handle cell weights.